mlrun/serving/server.py

```diff
@@ -526,6 +530,96 @@ def add_system_steps_to_graph(
     return graph
 
 
+async def _flush_all_batching_steps_single_pass(flow, logger, visited=None):
+    """
+    Flush all batching steps in a single pass through the graph.
+
+    Returns the total number of events flushed.
+    """
+    if visited is None:
+        visited = set()
+
+    events_flushed = 0
+
+    try:
+        # Walk through all steps in the flow
+        for step in flow._outlets:
+            # Skip if already visited (handles cyclic graphs)
+            if id(step) in visited:
+                continue
+            visited.add(id(step))
+
+            # Check if this step has batching capability
+            if hasattr(step, "_emit_all") and hasattr(step, "_batch"):
+                batch_count = sum(len(batch) for batch in step._batch.values())
+                if batch_count > 0:
+                    logger.info(
+                        f"Flushing {batch_count} buffered events from step '{step.name}'"
+                    )
+                    await step._emit_all()
+                    logger.info(f"Successfully flushed step '{step.name}'")
+                    events_flushed += batch_count
+
+            # Recursively flush child steps
+            if hasattr(step, "_outlets"):
+                events_flushed += await _flush_all_batching_steps_single_pass(
+                    step, logger, visited
+                )
+
+    except Exception as e:
+        logger.warning(f"Error flushing batching steps during drain: {e}")
+        # Don't fail the drain - better to lose some data than block rebalancing
```

Comment:
I'm not sure about this statement - especially as I'm afraid explicit-ack logic will be violated by this.